tpot/base.py

```diff
@@ -351,6 +336,35 @@ def __init__(self, generations=100, population_size=100, offspring_size=None,
         self._setup_toolbox()
 
 
+    def _setup_scoring_function(self, scoring):
+        if scoring:
+            if isinstance(scoring, str):
+                if scoring not in SCORERS:
+                    raise ValueError(
+                        'The scoring function {} is not available. Please '
+                        'choose a valid scoring function from the TPOT '
+                        'documentation.'.format(scoring)
+                    )
+            elif callable(scoring):
+                if not isinstance(scoring, _BaseScorer):
+                    # If the user passed a custom metric function, store it in the sklearn
+                    # SCORERS dictionary
+                    scoring_name = scoring.__name__
+                    greater_is_better = 'loss' not in scoring_name and 'error' not in scoring_name
+                    SCORERS[scoring_name] = make_scorer(scoring, greater_is_better=greater_is_better)
```

Comment:
Since `self.scoring_function` is now a string (name of one of the default scoring functions) OR a scoring function with the signature `my_scorer(estimator, X, y)`, perhaps we shouldn't store custom scoring functions in the `SCORERS` dictionary any more. We can store the result of this `make_scorer` call directly to `self.scoring_function`.

Reply:
Actually, we need to consider the broader use of the `SCORERS` dictionary. If someone passes a scoring function with the signature `my_scorer(estimator, X, y)`, I don't think it would work in the TPOT `score` function. Perhaps all custom scorers need to be stored in the `SCORERS` dictionary.

See [here](https://github.com/weixuanfu/tpot/blob/8252c742514f24bbe12e837534c0d21ea8caa09b/tpot/base.py#L843).

We should also have unit tests that cover all of these possible custom scorer cases.

Reply:
OK, will change it

Reply:
fixed in commit below. I think unit tests already cover all these scoring input types

Reply:
Looks the same to me; did the commit get pushed?

Reply:
Yep, check the comment below (or above)